cogboard-app/src/main/kotlin/com/cognifide/cogboard/logStorage/LogStorage.kt

```diff
@@ -0,0 +1,293 @@
+    private val logs: List<Log>
+    get() = logsCollection
+            ?.find()
+            ?.limit(config.logLines.toInt())
+            ?.sort(descending(Log.SEQ))
+            ?.mapNotNull { it }
+            ?.mapNotNull { Log.from(it) }
+            ?: emptyList()
+
+    /** Prepares a JSON response to be displayed */
+    private fun prepareResponse(): JsonObject {
+        return JsonObject(mapOf(
+            "variableFields" to parserStrategy.variableFields,
+            "logs" to JsonArray(logs.map { it.toJson() })
+        ))
+    }
+
+    /** Updates the logs and sends them to the widget. */
+    fun updateLogs(fetchNewLogs: Boolean) {
+        if (fetchNewLogs) {
+            // Download new logs
+            downloadLogs()
+
+            // Delete unnecessary logs
+            deleteOldLogs()
+            deleteSpaceConsumingLogs()
+        }
+
+        // Fetch the logs from the database and send them back
+        val response = prepareResponse()
+        vertx?.eventBus()?.send(config.eventBusAddress, response)
+    }
+
+    /** Deletes all data associated with the widget. */
+    fun delete() {
+        deleteAllLogs()
+        saveConfiguration(null)
+    }
+
+    companion object {
+        private const val DATABASE_NAME: String = "logs"
+        private const val CONFIGURATION_COLLECTION_NAME: String = "config"
+        private const val STATS_COMMAND: String = "collStats"
+        private const val STATS_SIZE: String = "size"
+        private const val MB_TO_BYTES: Long = 1024L * 1024L
+        private const val DAY_TO_TIMESTAMP = 24 * 60 * 60
+        private const val MONGO_SCHEME = "mongodb"
+        private val MONGO_USERNAME = System.getenv("MONGO_USERNAME") ?: "root"
+        private val MONGO_PASSWORD = System.getenv("MONGO_PASSWORD") ?: "root"
+        private val MONGO_HOST = System.getenv("MONGO_HOST") ?: "mongo"
+        private val MONGO_PORT = System.getenv("MONGO_PORT")?.toIntOrNull() ?: 27017
+        private var mongoClient: MongoClient? = null
+
+        /** Returns a shared instance of the Mongo client. */
+        private val client: MongoClient?
+        get() {
+            if (mongoClient != null) {
+                return mongoClient
+            }
+            try {
+                val uri = URI(
+                        MONGO_SCHEME,
+                        "$MONGO_USERNAME:$MONGO_PASSWORD",
+                        MONGO_HOST,
+                        MONGO_PORT,
+                        null,
+                        null,
+                        null
+                )
+                mongoClient = MongoClients.create(uri.toString())
+                return mongoClient
+            } catch (exception: MongoException) {
+                LOGGER.error("Cannot create a mongo client: $exception")
+            }
+            return null
+        }
+
+        /** Returns a database for storing logs and collection configurations. */
+        private val database: MongoDatabase?
```

Comment:
I can see many nullable fields like this. With Kotlin we should avoid this. Please make the code NPE safe where possible.

Reply:
Done.